common/web/lm-worker/src/main/model-compositor.ts

```diff
@@ -636,6 +565,92 @@ export default class ModelCompositor {
+   *
+   * Model-specific punctuation settings are applied during this process, as is
+   * verbose logging about the influence of "correction" vs "lexicon-frequency"
+   * per suggestion if enabled.
+   * @param deduplicatedSuggestionTuples The set of suggestions to finalize.
+   * @param context The context to which suggestions will be applied.
+   * @param inputTransform The actual text change to `context` that triggered
+   * the prediction request (which may be replaced by a Suggestion in the
+   * future)
+   * @returns
+   */
+  private finalizeSuggestions(deduplicatedSuggestionTuples: CorrectionPredictionTuple[], context: Context, inputTransform: Transform) {
+    const punctuation = this.punctuation;
+    const suggestions = deduplicatedSuggestionTuples.splice(0, ModelCompositor.MAX_SUGGESTIONS).map((tuple) => {
+      const prediction = tuple.prediction;
+
+      if(!this.verbose) {
+        return {
+          ...prediction.sample,
+          p: tuple.totalProb
+        };
+      } else {
+        const sample: Suggestion & {
+          p?: number,
+          "lexical-p"?: number,
+          "correction-p"?: number
+        } = {
+          ...prediction.sample,
+          p: tuple.totalProb,
+          "lexical-p": prediction.p,
+          "correction-p": tuple.correction.p
+        }
+
+        return sample;
+      }
+    });
+
+    // Apply 'after word' punctuation and casing (when applicable).  Also, set suggestion IDs.
+    // We delay until now so that utility functions relying on the unmodified Transform may execute properly.
+    suggestions.forEach((suggestion) => {
+      // Valid 'keep' suggestions may have zero length; we still need to evaluate the following code
+      // for such cases.
+
+      // If we're mid-word, delete its original post-caret text.
+      const tokenization = this.tokenize(context);
+      if(tokenization && tokenization.caretSplitsToken) {
+        // While we wait on the ability to provide a more 'ideal' solution, let's at least
+        // go with a more stable, if slightly less ideal, solution for now.
+        //
+        // A predictive text default (on iOS, at least) - immediately wordbreak
+        // on suggestions accepted mid-word.
+        suggestion.transform.insert += punctuation.insertAfterWord;
+
+        // Do we need to manipulate the suggestion's transform based on the current state of the context?
+      } else if(!context.right) {
+        suggestion.transform.insert += punctuation.insertAfterWord;
+      } else if(punctuation.insertAfterWord != '') {
+        if(context.right.indexOf(punctuation.insertAfterWord) != 0) {
+          suggestion.transform.insert += punctuation.insertAfterWord;
+        }
+      }
+
+      // If this is a suggestion after wordbreak input, make sure we preserve the wordbreak transform!
+      if(TransformUtils.isWhitespace(inputTransform)) {
+        let mergedTransform = models.buildMergedTransform(inputTransform, suggestion.transform);
```

Comment:
One really good thing that came out of this already:  the former `prefixTransform` variable wasn't particularly clear about its purpose.  It only existed to facilitate this check, and was only ever set by `TransformUtils.isWhitespace(inputTransform)`... which is considerably more clear to use directly here.

Reply:
It was also set if `allowSpace` is true:

https://github.com/keymanapp/keyman/pull/11899/files#diff-9ab3982966f4b8b331ddff4f5374a3b6ae038d44cd390f38c0dbe2598adf3275L180-L183

Reply:
> It was ~~also~~ set if `allowSpace` is true

https://github.com/keymanapp/keyman/blob/fccaf4eb61edd62525fb3154da5e0b486fd3a35d/common/web/lm-worker/src/main/model-compositor.ts#L151